sparkmagic/sparkmagic/kernels/wrapperkernel/sparkkernelbase.py

```diff
@@ -43,108 +49,168 @@ def __init__(self, implementation, implementation_version, language, language_ve
         # Disable warnings for test env in HDI
         requests.packages.urllib3.disable_warnings()
 
-        if not kwargs.get("testing", False):
-            self._load_magics_extension()
-            self._change_language()
-            if conf.use_auto_viz():
-                self._register_auto_viz()
```

Comment:
I don't see where all these functions are now being called from, and I'm guessing this is the reason magics don't get registered upon kernel startup.

Doesn't look like the `IPythonKernel` provides any `async` startup methods to call these from, so I see why you had to remove them...

I wonder if lazy initialization of sorts upon first call to `do_execute` would work here? Perhaps someone with more experience with IPython will suggest a better option.

Reply:
@sergiimk I tried

1. Lazily initializing the magics on the first call to `do_execute`
2. Running them synchronously within the `init` via `asyncio`

but neither approach works 😞 .  Let me know if you have any other thoughts.

Reply:
Sorry for late reply @devstein  - I was struggling to find time to take a closer look at the code.

Running load magics synchronously probably won't work:
- IO loops are non re-entrant, so you cannot block on a coroutine if the loop is already running
- If you create a new loop for those tasks it will have its own separate list of IO descriptors. So if it happens to await for anything that is managed by the original loop - it will block forever.

I tried lazy loading and ended up writing code very similar to your first attempt:
```python
    async def do_execute(...):
        if not self._lazy_initialized:
            self._lazy_initialized = True
            await self._load_magics_extension()
            await self._change_language()
            if conf.use_auto_viz():
                await self._register_auto_viz()
```
And it **mostly** worked in my notebooks - magics, auto viz was running as expected, but for some reason the images (loaded via markdown and via HTML) would not display. Only after killing the kernel a few times and switching back and forth to vanilla Python3 kernel the PySpark kernel would suddenly start displaying images. I suspect though that this is not related to this PR but a side effect of jupyter/notebook#6164.

Can you tell me what did not work for you when you tried lazy initializing the magics?